src/Workspaces/Core/Portable/Workspace/Solution/DocumentInfo.cs

```diff
@@ -71,22 +78,39 @@ internal DocumentInfo(DocumentAttributes attributes, TextLoader loader)
             string filePath = null,
             bool isGenerated = false)
         {
-            return new DocumentInfo(new DocumentAttributes(id, name, folders, sourceCodeKind, filePath, isGenerated), loader);
+            return Create(id, name, folders, sourceCodeKind, loader, filePath, isGenerated, documentServiceProvider: null);
+        }
+
+        internal static DocumentInfo Create(
+            DocumentId id,
+            string name,
+            IEnumerable<string> folders,
+            SourceCodeKind sourceCodeKind,
+            TextLoader loader,
+            string filePath,
+            bool isGenerated,
+            IDocumentServiceProvider documentServiceProvider)
+        {
+            return new DocumentInfo(new DocumentAttributes(id, name, folders, sourceCodeKind, filePath, isGenerated), loader, documentServiceProvider);
         }
 
         private DocumentInfo With(
             DocumentAttributes attributes = null,
-            Optional<TextLoader> loader = default)
+            Optional<TextLoader> loader = default,
+            IDocumentServiceProvider documentServiceProvider = null)
```

Comment:
why is this not an Optional<IDocumentServiceProvider>?

Reply:
null is not valid value. null means optional. at the end, if it is null, it will get default implementation.